### PR TITLE
feat(frontend): Complete Sub-DAG implementation in KFPv2.

### PR DESCRIPTION
**Pull request by zijianjoy**
**Description of your changes:**

1. Fix the problem that DAG status icon was not updated in real time.
1. Introduce `mlmdId` to the node, because there are multiple Executions/Artifacts with the same name, we need MLMD ID to identify runtime status within a MLMD object uniquely.
1. Make sure the Execution chosen for node's runtime status has a matching `parent_dag_id` with the `last element in breadcrumb layer's execution ID`.
2. Fix the styling of SubDAG status, so the status icon doesn't excess the border of Sub DAG.
3. Add a new layer for listing iterations in ParallelFor DAG. For example: ParallelFor `for-loop` has two iterations: `for-loop.0` and `for-loop.1`. Both iterations are rendered as Sub DAG nodes. And breadcrumb layer will show the iteration name if you open them.
4. If the ParallelFor DAG is not executed, we still show the SubDAG content by skipping iteration layer.
5. Render Artifact Node which is input of subDAG. Usually Artifact is an output of Execution. However, Artifact input of the current subDAG won't have an incoming edge. This PR draws Artifact input node and add an edges to connect it with downstream executions.
6. Instead of changing `flowElements` in place, `updateFlowElementsState()` returns an updated `flowElements` with runtime status.

Video:

https://user-images.githubusercontent.com/37026441/193442117-d076158a-1e6b-43e9-906c-caba4bcdec5f.mov

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: frontend/src/lib/v2/DynamicFlow.ts [hunk 1]
```diff
@@ -70,12 +216,19 @@ export const TASK_NAME_KEY = 'task_name';
...
 
 export function updateFlowElementsState(
+  layers: string[],
   elems: PipelineFlowElement[],
   executions: Execution[],
   events: Event[],
   artifacts: Artifact[],
-) {
+): PipelineFlowElement[] {
   // IMPORTANT: PipelineFlowElement update is in-place.
```
Comment (jlyaoyuli) [inline frontend/src/lib/v2/DynamicFlow.ts:225]: In the current design, the element update is longer in-place. Can you either remove the comment or change the comment? Thank you.

File: frontend/src/lib/v2/StaticFlow.ts [hunk 2]
```diff
@@ -271,19 +293,26 @@ function addArtifactToTaskEdges(
...
+        flowGraph.push(edge);
+      } else if (componentInputArtifact) {
+        const edge: Edge = {
+          id: getArtifactToTaskEdgeKey(componentInputArtifact, inputTaskKey),
+          source: getArtifactNodeKey('', componentInputArtifact),
+          target: getTaskNodeKey(inputTaskKey),
+          arrowHeadType: ArrowHeadType.ArrowClosed,
+        };
+        flowGraph.push(edge);
       }
```
Comment (jlyaoyuli) [inline frontend/src/lib/v2/StaticFlow.ts:315]: Are these two conditions also handle the `input artifact -> subDAG` and `subDAG -> output artifact` ? 

Screen Shot 2022-10-10 at 10 06 40 AM

File: frontend/src/lib/v2/DynamicFlow.ts [hunk 3]
```diff
@@ -136,23 +340,28 @@ export function getNodeMlmdInfo(
...
+      .get(taskName)
+      ?.filter(exec => exec.getId() === elem.data?.mlmdId);
+    return executions ? { execution: executions[0] } : {};
   }
   return {};
 }
 
-function getTaskNameToExecution(executions: Execution[]): Map<string, Execution> {
-  const map = new Map<string, Execution>();
+function getTaskNameToExecution(executions: Execution[]): Map<string, Execution[]> {
```
Comment (jlyaoyuli) [inline frontend/src/lib/v2/DynamicFlow.ts:351]: (Correct me if I am wrong.) I think the function name `getTaskNameToExecution` is a little confusing especially we are trying do some **set** works. Would it be better if we change the name to `setTaskNameToExecution`? If you think the updated function name is better, can you also change the names for other functions?

File: frontend/src/lib/v2/DynamicFlow.ts (hunk 1 above)
Comment (zijianjoy) [reply frontend/src/lib/v2/DynamicFlow.ts:225]: Good point! Removed the comment.

File: frontend/src/lib/v2/StaticFlow.ts (hunk 2 above)
Comment (zijianjoy) [reply frontend/src/lib/v2/StaticFlow.ts:315]: `subDAG -> output artifact` is able to be handled already before this PR. `input artifact -> subDAG` support is added in this PR.

File: frontend/src/lib/v2/DynamicFlow.ts (hunk 3 above)
Comment (zijianjoy) [reply frontend/src/lib/v2/DynamicFlow.ts:351]: `setXXX()` is a naming pattern for changing existing object, which means `executions: Execution[]` parameter in this case. However, we are returning `Map<string, Execution[]>` which is the function caller tries to *get*, so I use `getXXX()` pattern here.

**Comment by jlyaoyuli** [comment]
/lgtm
/approve
Thank you @zijianjoy for implementing the sub-DAG feature!

**Pull request merged**